tests-wasm/y-doc.tests.js

```diff
@@ -265,4 +265,31 @@ export const testLiveness = tc => {
         t.assert(!a2.alive(tx), 'child is deleted (local)')
         t.assert(!aa2.alive(tx), 'parent was deleted (local)')
     })
+}
+
+/**
+ * @param {t.TestCase} tc
+ */
+export const testRoots = tc => {
+    const d1 = new Y.YDoc()
+    const a = d1.getMap('a')
+    const b = d1.getText('b')
+    const c = d1.getArray('c')
+    const d = d1.getXmlFragment('d')
+
+    const roots = d1.roots()
+        .sort((a, b) => a[0].localeCompare(b[0]))
+        .map(([k,v]) => [k, v.constructor])
+    t.compare(roots, [
+        ['a', Y.YMap],
+        ['b', Y.YText],
+        ['c', Y.YArray],
+        ['d', Y.YXmlFragment]
+    ])
+
+    a.set('hello', 'world')
+
+    let d2 = new Y.YDoc()
+    exchangeUpdates([d1, d2])
+    t.compare(d2.roots(), [['a', undefined]])
```

Comment:
You can get the raw branch, but these are not exposed outside. We also don't know what to map them to, as for root types, we don't pass their type (YText, YMap etc.) as part of the update payload. Since you have a name, you can technically map them onto anything you want by calling `doc.getText(name)`, `doc.getMap(name)` etc.

Reply:
> for root types, we don't pass their type (YText, YMap etc.) as part of the update payload

Maybe types should be passed in the update payload? Otherwise it feels like we are loosing information, as we cannot totally sync with remote data if we don't have prior information about what we are going to get.
In pycrdt the document is really seen as a map which can be iterated on, and ideally a document structure should be completely discoverable. For instance, I'd like to inspect the type of each entry and do something accordingly:

```py
doc = Doc()
doc.apply_update(update)
for key, value in doc.items():
    if isinstance(value, Text):
        ...
    elif isinstance(value, Map):
        ...
```

Reply:
In general root level types are considered to be static and well known - kinda like tables are in the SQL databases. If you need an iterable / discoverable structure that's a subject of some conditional logic, just create a single root level map, then push other types (that you'd normally consider root level) as its children.

Reply:
Indeed, which raises the question of the utility of root types. Why couldn't we just have a (hidden) YMap at the root level? The problem with root types is that we say they "should not be changed once created", yet we allow casting them to another type, and they are not fully discoverable.

Reply:
@davidbrochart If that is what you want, you can already create a root Y.Map and only use that.

However, then clients will overwrite each other's content whenever they create a new property on that root-level Y.Map. Subsequent calls of `ymap.set('content', new Y.Text())` will overwrite the previous value. So every reload of a page will result in a modified document. We need a way to agree that there are certain root-types that should not be overwritten. E.g. `content` is always a Y.Text.

This is why clients should agree beforehand on what the structure of the Y.Doc should look like. E.g `ydoc.get('content')` is already a `Y.Text`.

Usually, you would solve this using some kind of `schema`. However, schemas in CRDTs are infeasible (I've never seen an approach with favorable tradeoffs). IMHO a dynamic schema, as it is implemented right now, suits the nature of CRDTs much better (they require no central coordination service that can enforce a specific schema). Though I agree, it's unintuitive if you come from a database perspective.

Reply:
I'd be happy to discuss why you think schemas in CRDTs are infeasible, because this is something I'm looking at currently in pycrdt. Because pycrdt's primary use-case is in a server, it is important to accept changes to a document that are valid (respect a defined schema) and authorized (this client cannot change this attribute).
Also, I don't understand what a dynamic schema is. Imagine a peer changes a root type from a YText to a YMap, I don't even see how you can cast the old text to the new map. For me it means that the old value is lost, and since the other peers are not notified of the type change, how is this supposed to work? They will remain operating on what they believe is a text while the other peer will operate on it as a map. I'm very confused on how this can be used.

Reply:
A schema is definitely possible in an architecture that has some kind of "source of truth" (a trusted server that is responsible for handling connections). 

However, one appeal of CRDTs is that they don't require a source of truth. This makes it, for example, easier to scale CRDTs for more devices than a single server could handle.

If there is no authority / source of truth, then there can be no single schema that everyone is forced to use. 

Every client is responsible for making sense of the data.

If user A interprets root-type "data" as a Y.Map, then they will see all key-value pairs stored on that root-type.

If user B interprets root-type "data" as a Y.Array, then they will see all items stored in the list property of that root-type.

If user C interprets root-type "data" as a Y.XmlElement, then will see an XML node that contains children (the items of the Y.Array) and it will have properties (the properties stored on the Y.Map).

Users are responsible for interpreting the data. And custom data types can help you to make sense of evolving (or even diverging) schemas. However, that is currently out-of-scope.

That said, I still belive that the user of the YCRDT is responsible for enforcing some kind of schema (at the least, make sure that it is forward-compatible). I don't believe that we can build a schema approach that will work for all users of Yjs & Y-CRDT.





Something that works quite well in Yjs is that instead of defining a schema (e.g. root-type "data" must be a Y.Map), everyone can interpret the data.

Reply:
Thanks for the details!